Move get_frame_string from Copter to AP_Motors

kd0aij opened the pull request:
first 2 commits move the copter-only frame_string enum to AP_Motors_class.h
and add a new static string method get_frame_type().
The first plane commit switches to using the class and type strings.
Second plane commit changes from gcs().send_text to hal.console->printf since the gcs never receives those messages. I was using sitl/mavproxy to test this, but perhaps I was doing something wrong?
 Third plane commit adds a log message for quadplane which is similar to copter's.

khancyr reviewed (commented):
I don't think th gcs ->printf are right, did you check by issuing a reboot on SITL? You could miss some startup message launching SITL with sim_vehicle.py

kd0aij commented:
they don't show up on a reboot in SITL either

peterbarker commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -64,6 +101,47 @@ class AP_Motors {
+    {
+        switch((AP_Motors::motor_frame_type)frame_type) {
+            case MOTOR_FRAME_TYPE_PLUS:
+                return "PLUS";
```
These strings should exist on the classes themselves, rather than in a switch statement here.  So:

```
pbarker@bluebottle:~/rc/ardupilot(pr/decode-watchdog)$ mavproxy.py --help | lesspbarker@bluebottle:~/rc/ardupilot(pr/decode-watchdog)$ git diff
diff --git a/libraries/AP_Motors/AP_MotorsSingle.h b/libraries/AP_Motors/AP_MotorsSingle.h
index 5276c25d74..cadf9e05b6 100644
--- a/libraries/AP_Motors/AP_MotorsSingle.h
+++ b/libraries/AP_Motors/AP_MotorsSingle.h
@@ -31,6 +31,8 @@ public:
     // init
     void                init(motor_frame_class frame_class, motor_frame_type frame_type) override;
 
+    const char *name() const override { return "SINGLE"; }
+
     // set frame class (i.e. quad, hexa, heli) and type (i.e. x, plus)
     void set_frame_class_and_type(motor_frame_class frame_class, motor_frame_type frame_type) override;
 
pbarker@bluebottle:~/rc/ardupilot(pr/decode-watchdog)$ 
```

kd0aij replied:
sorry, I don't understand; frame_type strings should be handled differently than frame_class?

peterbarker replied:
Both should be retrievable in the same way - as method calls on the motors class.

Presumably MotorsMatrix would need to switch() on `_last_frame_class` and `_last_frame_type`

I'm guessing that `frame_name()` might return nullptr for in the base class?

kd0aij replied:
I see; didn't notice that frame_class and type were attributes.

kd0aij replied:
why not move _last_frame_class and _last_frame_type up to the base class and require init() and set_frame_class_and_type() to maintain them?

peterbarker replied:
Try it and see?  I think the answer might be, "because only MotorsMatrix cares about frame type" - but that should become evident.

Most of the classes should be returning a static string for the class - they are what they are, and we shouldn't change the string being returned from motors based on the current configuration, rather returning what's allocated and/or active.

peterbarker commented on ArduPlane/Log.cpp:
```diff
@@ -511,6 +511,10 @@ void Plane::Log_Write_Vehicle_Startup_Messages()
     // only 200(?) bytes are guaranteed by AP_Logger
     Log_Write_Startup(TYPE_GROUNDSTART_MSG);
+    if (quadplane.initialised) {
+        logger.Write_MessageF("QuadPlane Frame: %s, Type: %s", AP_Motors::get_frame_string(quadplane.frame_class.get()),
```
This would call methods on the allocated motors backend.

kd0aij commented:
reworked

also adds frame_type strings

moves the get_frame_string method into the AP_Motors base class along with _last_frame_class and type, renaming them to _active_frame_class/type
adds a new non-virtual base class method initialise(class, type) which sets the active frame class/type and then calls the original pure virtual init(class, type) 
Since AP_Motors6DOF inherits from AP_MotorsMatrix, but uses a different frame class enum (sub_frame_t), for correctness the get_frame_string method is virtual and overridden by AP_Motors6DOF, although this method is currently unused by Sub.
When testing this, I noticed that copter tailsitter setup in quadplane was susceptible to configuration errors involving frame classes not supported by AP_MotorsMatrix;  these now result in a call to config_error.
Also, the gcs().send_text calls in QuadPlane::setup occur before the GCS is connected; changed these to ::printf since they were apparently of no value to the user.
Also added a quadplane frame class/type message to the log similar to Copter's.

peterbarker commented:
@kd0aij did you mean to close this?

kd0aij commented:
yeah, screwed up trying to push a new branch over it
just created a new one #15748
